### ISSUE TITLE
Store image and metadata on chain

### ISSUE THREAD

**Issue by vpontis** [glow-xyz/nftoken#18]
Inspired by discussion in #17

Some users may want to have _more_ info on chain. 

- Chain Runners is an NFT project where the art is on chain
- Some NFT people have talked about doing metadata storage on chain. That would enable things like placing a bid on a group of traits. 

To do this, we could hack the `image_url` / `metadata_url` fields to support addresses by saying they are an onchain address if they start with `s:`.

**Comment by vovacodes** [comment]
What abouta a schema for metadata? Is it opaque and left for dapps to serialize/deserialize according to their own needs? I feel like it can be quite neat to allow creating a schema PDA that can be linked to the metadata so that every dapp/wallet would know how to parse the data. WDYT?

**Comment by vpontis** [comment]
Yea, there could be a schema for Metadata but I think most of the fields should be optional.

**Comment by jpbogle** [comment]
I agree something in the form of a schema. I have thought about this a lot and played around with on-chain NFTs a good amount already. Currently we use a serverless function which uses the anchor IDL schemas and then a way to define fields to pull from that schema and assign human names to each field, and returns that in the metaplex json format as attributes. We have a program that stores this schema but arent using it (https://github.com/cardinal-labs/cardinal-generator) in any production use case yet

I was originally thinking something where the uri contains both the address to the account, and address to the schema. Something like `s://${account}.${schema}` and this schema basically contains array of tuples containing the address to lookup, the byteoffset for that attribute, type and name

However, I think what you actually want and is a lot more powerful is to allow this schema to point to many attributes where each attribute can come from different accounts. This allows for composition at the attribute level so an on-chain NFT could aggregate data across multiple programs into attributes. So I think when defining an on-chain NFT it could point to a schema program that just describes the base URI, along with a vector of attribute, where each attribute describes where to find it, how to deserialize and display it. Something like

```
#[derive(Default, Clone, AnchorSerialize, AnchorDeserialize)]
pub struct Attribute {
    pub address: Pubkey,
    pub byte_offset: u64,
    pub attribute_type: String,
    pub name: String,
    pub display_name: Option<String>,
}

#[account]
pub struct MetadataConfig {
    pub base_metadata_uri: String,
    pub attributes: Vec<Attribute>,
}
```

Another option could be instead of storing the schema address in the URI, having it derived in a common way and stored directly in this nftoken program or a separate nftoken metadata program if we think it would clog getProgramAccounts calls and indexing. So everyone knows to look at this program, and the PDA derived from ["schema", nft address] and you can just add this additionally and someone displaying this NFT can choose to display it / merge it with the existing URI field in the nft token.

I think I am in favor of setting it in the URI since it forces someone trying to display this NFT to support this schema type, rather than just choosing to not honor the on-chain attributes, but maybe just putting in a common place would work. I would assume beyond attributes there is still value to off-chain data as well so it could just also store a base_uri field but maybe thats not necessary

**Comment by vpontis** [comment]
@jpbogle thanks for writing this up. I think I did not totally follow all of your reasoning or your conclusion. 

What is the benefit of having the Metadata split across multiple accounts rather than stored in a single separate account?

**Comment by vpontis** [comment]
@jpbogle thank you for your examples. That makes it much more clear what you're thinking about. 

So I think the way this could work is that you still have one account that defines the metadata which is probably a list of (nested) key pairs, but then for the traits you could do something like:

- `gas_tank_level → s://account/u64/name`

And then the client can look up the current `gas_tank_level` or something like that.
